File: gtci/01-two-pointers/python3/6-triplet-with-smaller-sum.py
class Solution:
    def searchTriplets(self, arr, target):
        count = 0

        arr.sort()

        for idx, num in enumerate(arr):
            left = idx + 1
            right = len(arr) - 1

            while left < right:
                current_sum = num + arr[left] + arr[right]

                if current_sum < target:
                    count += right - left
                    left += 1
                elif current_sum >= target:
                    right -= 1

        return count


if __name__ == "__main__":
    sol = Solution()

    assert sol.searchTriplets([-2, 0, 1, 3], 2) == 2
    assert sol.searchTriplets([-1, 4, 2, 1, 3], 5) == 4
    assert sol.searchTriplets([0, 0, 0, 0, 0], 1) == 10
    print("All test cases passed.")

File: gtci/01-two-pointers/cpp/6-triplet-with-smaller-sum.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
   public:
    int searchTriplets(vector<int> &&arr, int target) {
        int count = 0;

        sort(arr.begin(), arr.end());

        for (int i = 0; i < arr.size(); i++) {
            const int firstNum = arr[i];

            vector<int>::const_iterator leftItr = arr.cbegin() + i + 1;
            vector<int>::const_iterator rightItr = arr.cend() - 1;

            while (leftItr < rightItr) {
                const int currentSum = firstNum + *leftItr + *rightItr;

                if (currentSum < target) {
                    count += rightItr - leftItr;
                    leftItr++;
                } else {
                    rightItr--;
                }
            }
        }

        return count;
    }
};

int main() {
    Solution sol;

    assert(sol.searchTriplets(vector<int>{-2, 0, 1, 3}, 2) == 2);
    assert(sol.searchTriplets(vector<int>{-1, 4, 2, 1, 3}, 5) == 4);
    assert(sol.searchTriplets(vector<int>{0, 0, 0, 0, 0}, 1) == 10);
    cout << "All test cases passed." << endl;

    return 0;
}
